src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -325,6 +356,459 @@ CreateInsertSelectRouterPlan(Query *originalQuery,
+													 eqClass->equivalenceId);
+
+				/*
+				 * It seems inefficient to start from the beginning.
+				 * But, we should somehow restart from the beginning to test that
+				 * whether the already skipped ones are equal or not.
+				 */
+				equivalenceClassIndex = 0;
+
+				break;
+			}
+		}
+	}
+
+	return mainEquivalenceClass;
+}
+
+
+/*
+ * Per each relation restriction. get the eq_classes that is in the plannerInfo.
+ * Then, add all the equivalent members of a class to a var equivalence class.
+ */
+static List *
+GenerateVarEquivalencesForRelationRestrictions(
+	RelationRestrictionContext *restrictionContext)
+{
+	List *varEquivalences = NIL;
+	ListCell *relationRestrictionCell = NULL;
+
+	foreach(relationRestrictionCell, restrictionContext->relationRestrictionList)
+	{
+		RelationRestriction *relationRestriction = lfirst(relationRestrictionCell);
+		List *equivalenceClasses = relationRestriction->plannerInfo->eq_classes;
+		ListCell *equivilanceClassCell = NULL;
+		PlannerInfo *plannerInfo = relationRestriction->plannerInfo;
+
+		foreach(equivilanceClassCell, equivalenceClasses)
+		{
+			EquivalenceClass *plannerEqClass = lfirst(equivilanceClassCell);
+			ListCell *equivilanceMemberCell = NULL;
+			VarEquivalenceClass *varEquivalance = palloc0(sizeof(VarEquivalenceClass));
+			varEquivalance->equivalenceId = varEquivalenceId++;
+
+			foreach(equivilanceMemberCell, plannerEqClass->ec_members)
+			{
+				EquivalenceMember *em = lfirst(equivilanceMemberCell);
+				Var *expressionVar = NULL;
+
+				if (!IsA(em->em_expr, Var))
+				{
+					continue;
+				}
+
+				expressionVar = (Var *) em->em_expr;
+				AddToVarEquivalenceClass(plannerInfo, expressionVar, &varEquivalance);
+			}
+
+			varEquivalences = lappend(varEquivalences, varEquivalance);
+		}
+	}
+
+	return varEquivalences;
+}
+
+
+/*
+ * We now added a join restriction that keeps track of non anti join restrictions.
+ * For those joins, use the generated restriction information. Then, add all the
+ * join restriction information members of a class to a var equivalence class.
+ */
+static List *
+GenerateVarEquivalencesForJoinRestrictions(JoinRestrictionContext *joinRestrictionContext)
+{
+	List *varEquivalences = NIL;
+	ListCell *joinRestrictionCell = NULL;
+
+	foreach(joinRestrictionCell, joinRestrictionContext->joinRestrictionList)
+	{
+		JoinRestriction *joinRestriction = lfirst(joinRestrictionCell);
+		ListCell *restrictionInfoList = NULL;
+
+		VarEquivalenceClass *varEquivalance = palloc0(sizeof(VarEquivalenceClass));
+		varEquivalance->equivalenceId = varEquivalenceId++;
+
+		foreach(restrictionInfoList, joinRestriction->joinRestrictInfoList)
+		{
+			RestrictInfo *rinfo = (RestrictInfo *) lfirst(restrictionInfoList);
+			OpExpr *restrictionOpExpr = NULL;
+			Expr *leftExpr = NULL;
+			Expr *rightExpr = NULL;
+			Var *leftVar = NULL;
+			Var *rightVar = NULL;
+			Expr *restrictionClause = rinfo->clause;
+
+			if (!IsA(restrictionClause, OpExpr))
```

Comment:
> ON (r1.a = r1.b and false)

For this one, Postgres understands that the join wouldn't yield and rows thus doesn't generate any join  restrictions. In the later planning, both relations' restriction information includes `false` and thus all shards pruned.

Example:
```SQL

INSERT INTO agg_events                                                                                                                                                                                                                                      (value_1_agg,                                                                                                                                                                                                                                               user_id)                                                                                                                                                                                                                                      SELECT SUM(value_1),                                                                                                                                                                                                                                              id                                                                                                                                                                                                                                                  FROM   (SELECT raw_events_second.user_id AS id,                                                                                                                                                                                                                           raw_events_second.value_1                                                                                                                                                                                                                           FROM   raw_events_first,                                                                                                                                                                                                                                          raw_events_second                                                                                                                                                                                                                                   WHERE  raw_events_first.value_1 = 15 AND raw_events_first.user_id = raw_events_second.user_id  and false) AS foo                                                                                                                                   GROUP  BY id;
DEBUG:  StartTransactionCommand
DEBUG:  StartTransaction
DEBUG:  name: unnamed; blockState:       DEFAULT; state: INPROGR, xid/subid/cid: 0/1/0, nestlvl: 1, children: 
DEBUG:  Skipping target shard interval 102072 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102073 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102074 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102075 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102076 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102077 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102078 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102079 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102080 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102081 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102082 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102083 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102084 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102085 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102086 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102087 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102088 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102089 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102090 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102091 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102092 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102093 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102094 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102095 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102096 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102097 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102098 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102099 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102100 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102101 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102102 since SELECT query for it pruned away
DEBUG:  Skipping target shard interval 102103 since SELECT query for it pruned away
DEBUG:  ProcessQuery
DEBUG:  Plan is router executable
DEBUG:  CommitTransactionCommand
DEBUG:  CommitTransaction
DEBUG:  name: unnamed; blockState:       STARTED; state: INPROGR, xid/subid/cid: 0/1/0, nestlvl: 1, children: 
INSERT 0 0
```

> ON (r1.a =  r2.b or r1.c = NULL)

For this one, good catch I haven't considered this at all. 

The restriction is generated to be `BoolExpr` where the logic here expects `OpExpr` and thus not able to push the query. 

But, as far as I can see, we shouldn't allow OR clauses. So, the behavior seems OK to me. What do you think?